fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,666 @@
+
+/**
+ * Main FUSE implementation class.
+ *
+ * Implements the FUSE callbacks defined by jnr-fuse.
+ */
+final class TachyonFuseFs extends FuseStubFS {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final int MAX_OPEN_FILES = Integer.MAX_VALUE;
+  private static final long[] UID_AND_GID = TachyonFuseUtils.getUidAndGid();
+
+  private final TachyonConf mTachyonConf;
+  private final TachyonFileSystem mTFS;
+  // base path within Tachyon namespace that is used for FUSE operations
+  // For example, if tachyon-fuse is mounted in /mnt/tachyon and mTachyonRootPath
+  // is /users/foo, then an operation on /mnt/tachyon/bar will be translated on
+  // an action on the URI tachyon://<master>:<port>/users/foo/bar
+  private final Path mTachyonRootPath;
+  private final String mTachyonMaster;
+  // Keeps a cache of the most recently translated paths from String to TachyonURI
+  private final LoadingCache<String, TachyonURI> mPathResolverCache;
+
+  // Table of open files with corresponding InputStreams and OutptuStreams
+  private final Map<Long, OpenFileEntry> mOpenFiles;
+  private long mNextOpenFileId;
+
+  TachyonFuseFs(TachyonConf conf, TachyonFileSystem tfs, TachyonFuseOptions opts) {
+    super();
+    mTachyonConf = conf;
+    mTFS = tfs;
+    mTachyonMaster = mTachyonConf.get(Constants.MASTER_ADDRESS);
+    mTachyonRootPath = Paths.get(opts.getTachyonRoot());
+    mNextOpenFileId = 0L;
+    mOpenFiles = Maps.newHashMap();
+
+    final int maxCachedPaths = mTachyonConf.getInt(Constants.FUSE_PATHCACHE_SIZE);
+    mPathResolverCache = CacheBuilder.newBuilder()
+        .maximumSize(maxCachedPaths)
+        .build(new PathCacheLoader());
+
+    Preconditions.checkArgument(mTachyonRootPath.isAbsolute(),
+        "tachyon root path should be absolute");
+  }
+
+  /**
+   * Creates and opens a new file.
+   * @param path The FS path of the file to open
+   * @param mode mode flags
+   * @param fi FileInfo data struct kept by FUSE
+   * @return 0 on success. A negative value on error
+   */
+  @Override
+  public int create(String path, @mode_t long mode, FuseFileInfo fi) {
+    // mode is ignored in tachyon-fuse
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    final int flags = fi.flags.get();
+    LOG.trace("create({}, {}) [Tachyon: {}]", path, Integer.toHexString(flags), turi);
+    final int openFlag = flags & 3;
+    if (openFlag != O_WRONLY.intValue()) {
+      OpenFlags flag = OpenFlags.valueOf(openFlag);
+      LOG.error("Passed a {} flag to create(). Files can only be created in O_WRONLY mode ({})",
+          flag.toString(), path);
+      return -ErrorCodes.EACCES();
+    }
+
+    try {
+
+      final OpenFileEntry ofe =
+          new OpenFileEntry(null, mTFS.getOutStream(turi));
+      LOG.debug("Tachyon OutStream created for {}", path);
+
+      synchronized (mOpenFiles) {
+        if (mOpenFiles.size() == MAX_OPEN_FILES) {
```

Comment:
NIT: `mOpenFiles.size() >= MAX_OPEN_FILES` to make the code more robust against bugs?